test/end2end_test.go

```diff
@@ -6880,6 +6881,174 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for {
+				_, err := stream.Recv()
+				if err == io.EOF {
+					return nil
+				}
+				if err != nil {
+					return err
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return err
+				}
+			}
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("Starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	var streams []streamWithPeer
+
+	// Setting up 3 streams and calling Send() and Recv() once on each.
+	for i := 0; i < 3; i++ {
+		p := new(peer.Peer)
+		s, err := ss.Client.FullDuplexCall(ctx, grpc.Peer(p))
+		if err != nil {
+			t.Fatalf("Creating FullDuplex stream: %v", err)
+		}
+		sp := streamWithPeer{stream: s, peer: p, id: "stream " + strconv.Itoa(i)}
+		streams = append(streams, sp)
+
+		if err := sp.sendAndReceive(); err != nil {
+			t.Fatal(err)
+		}
+	}
+	// Verifying only 2 connections are created so far.
+	if len(creds.connections) != 2 {
+		t.Fatalf("Number of connections created: %v, want: 2", len(creds.connections))
+	}
+
+	// Verifying that streams on the first conn still works.
+	if err := streams[0].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
+	if err := streams[1].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
+
+	// The peer passed via the call option is set up only after the RPC is complete,
+	// which is why we wait here until Recv() returns EOF.
+	for i := 0; i < 3; i++ {
+		if err := streams[i].stream.CloseSend(); err != nil {
+			t.Fatalf("CloseSend() on %s: %v", streams[i].id, err)
+		}
+		for {
+			_, err := streams[i].stream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				t.Fatalf("Receiving on %s got: %v, want: EOF", streams[i].id, err)
+			}
+		}
+		streams[i].conn = streams[i].peer.AuthInfo.(*authInfoWithConn).connection
+	}
+
+	// Verifying the first and second RPCs were made on the same connection.
+	if streams[0].conn != streams[1].conn {
+		t.Fatal("Got streams using different connections; want same.")
+	}
+	// Verifying the third and first/second RPCs were made on different connections.
+	if streams[2].conn == streams[0].conn {
+		t.Fatal("Got streams using same connections; want different.")
+	}
+
+	// Verifying first connection was closed.
+	if _, err := creds.connections[0].closeCh.Receive(ctx); err != nil {
+		t.Fatal("Timeout expired when waiting for first client transport to close")
+	}
+}
```

Comment:
Can you move this test out of `end2end_test.go` please?

Maybe a new `transport_test.go` if there isn't something existing that works well?

Also, pro tip: copy/paste the `import` section from the top of this file into the new file and when you run `goimports` (on save in your IDE?) it will only need to prune the unused imports instead of try to determine what all the undefined symbols are and where to import them from.

Reply:
Sure. Ill move just this test for now. But ill try moving more transport tests out from here as part of #2330